day3/schematic.go

```diff
@@ -0,0 +1,228 @@
+// in the file at the specified path.
+func MakeSchematic(path string) (*Schematic, error) {
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	data := strings.Split(string(content), "\n")
+	return &Schematic{Data: data}, nil
+}
+
+func (schematic *Schematic) FindAllParts() ([][]PartNumber, error) {
+	partNumbers := make([][]PartNumber, len(schematic.Data))
+
+	// find all part numbers
+	for i, line := range schematic.Data {
+		chars := strings.Split(line, "")
+		for j := 0; j < len(chars); j++ {
+			_, err := strconv.Atoi(chars[j])
+			if err != nil {
+				continue
+			}
+			start := j
+
+			for err == nil {
+				j++
+				if j >= len(chars) {
+					break
+				}
+				_, err = strconv.Atoi(chars[j])
+			}
+			end := j - 1
+			number, _ := strconv.Atoi(line[start : end+1])
+			partNumbers[i] = append(partNumbers[i], PartNumber{
+				Number:   number,
+				StartIdx: start,
+				EndIdx:   end,
+			})
+		}
+	}
+
+	regex, _ := regexp.Compile(`[^0-9\.]`)
+
+	// validate all part numbers against schematic
+	for i, line := range schematic.Data {
+		chars := strings.Split(line, "")
+		for j, char := range chars {
+			match := regex.MatchString(char)
+			if !match {
+				continue
+			}
+			// check PNs above
+			for k := 0; k < len(partNumbers[i-1]); k++ {
+				if partNumbers[i-1][k].IsValid {
+					continue
+				}
+				if partNumbers[i-1][k].StartIdx <= j-1 && j-1 <= partNumbers[i-1][k].EndIdx {
+					partNumbers[i-1][k].IsValid = true
+				} else if partNumbers[i-1][k].StartIdx <= j && j <= partNumbers[i-1][k].EndIdx {
+					partNumbers[i-1][k].IsValid = true
+				} else if partNumbers[i-1][k].StartIdx <= j+1 && j+1 <= partNumbers[i-1][k].EndIdx {
+					partNumbers[i-1][k].IsValid = true
+				}
+			}
+			// check left and right
+			for k := 0; k < len(partNumbers[i]); k++ {
+				if partNumbers[i][k].IsValid {
+					continue
+				}
+				if partNumbers[i][k].EndIdx == j-1 || partNumbers[i][k].StartIdx == j+1 {
+					partNumbers[i][k].IsValid = true
+				}
+			}
+			// check PNs below
+			for k := 0; k < len(partNumbers[i+1]); k++ {
+				if partNumbers[i+1][k].IsValid {
+					continue
+				}
+				if partNumbers[i+1][k].StartIdx <= j-1 && j-1 <= partNumbers[i+1][k].EndIdx {
+					partNumbers[i+1][k].IsValid = true
+				} else if partNumbers[i+1][k].StartIdx <= j && j <= partNumbers[i+1][k].EndIdx {
+					partNumbers[i+1][k].IsValid = true
+				} else if partNumbers[i+1][k].StartIdx <= j+1 && j+1 <= partNumbers[i+1][k].EndIdx {
+					partNumbers[i+1][k].IsValid = true
+				}
+			}
+		}
+	}
+
+	return partNumbers, nil
+}
+
+// MakeAsteriskMap returns a map representing the surrounding area of an asterisk in a grid.
+// The asterisk is located at the specified line and linePosition.
+// The map contains boolean values indicating whether each position needs to be assessed
+// (i.e. 'false' indicates that this location can be skipped).
+func (schematic *Schematic) MakeAsteriskMap(line int, linePosition int) map[int]map[int]bool {
```

Comment:
This is the structure which simultaneously provides the path for walking around the asterisk symbol, as well as memorizing any points which have been visited already (indicated by a `false` value, meaning that the position does NOT need to be evaluated). I actually use this to throw false values in applicable areas for the situations when an asterisk is found on an edge in the schematic (top, bottom, left, or right).